src/System.Private.ServiceModel/src/System/ServiceModel/Channels/MessageContent.cs

```diff
@@ -0,0 +1,266 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.Diagnostics.Contracts;
+using System.Globalization;
+using System.IO;
+using System.Net;
+using System.Net.Http;
+using System.Net.Http.Headers;
+using System.Runtime;
+using System.Threading.Tasks;
+using System.Linq;
+
+namespace System.ServiceModel.Channels
+{
+    internal abstract class MessageContent : HttpContent
+    {
+        protected Message _message;
+        protected MessageEncoder _messageEncoder;
+        protected Stream _stream = null;
+        private bool _disposed;
+
+        public MessageContent(Message message, MessageEncoder messageEncoder)
+        {
+            _message = message;
+            _messageEncoder = messageEncoder;
+
+            SetContentType(_messageEncoder.ContentType);
+            PrepareContentHeaders();
+        }
+
+        public Message Message
+        {
+            get
+            {
+                return _message;
+            }
+        }
+
+        private void PrepareContentHeaders()
```

Comment:
This is the code from HttpInput that set headers on the HttpContent. As we have the message available to us, the HttpContent class should be the one to set the properties.